### DISCUSSION TITLE
Generic type/function parameters over values

### DISCUSSION THREAD

**Discussion by Shad0wRim** [gleam-lang/gleam#3550] [Ideas & suggestions]
In Rust, there are const generics that allow a type to be generic over a value rather than a type:
```Rust
fn foo<const N: usize>(arr: [i32; N]) { todo!(); }

struct Bar<const N: usize>([i32; N])
```

I would love something like this to be put into gleam, using perhaps using `const` like rust or just by specifying the type of the parameter in the type parameter.

```gleam
// rust-like
type MyType(const n: Int) {
  Phantom(a: Int, b: Int)
  Concrete(a: Int, n: Int)
}
// without const
type MyType(n: Int) {
  Phantom(a: Int, b: Int)
  Concrete(a: Int, b: Int, n: Int)
}
```

Instantiating this type would require a type hint, as I am unsure how it could be specified in the constructor if it is used as a "phantom" parameter. 
Perhaps if the type parameter name shares the same name and type as one of the fields in the record, the type hint could be removed, but the corresponding field would have to be known at compile time.

```gleam

fn main() {
  // type hinted
  let val: MyType(5) = Phantom(1, 2)
  let defined = Concrete(1, 2, 5)
}
```

Functions that use this type could extract the const type parameter with similar syntax to current generics

```gleam
fn my_func(val: MyType(n)) {
  let assert Phantom(a, b) = val
  a * n
}
```

Would love some feedback on this, and would love something like this as I ran into an issue where I couldn't create a type representing the rationals adjunct square_root(n) without creating a separate type for each n or supplying the n to each function call, making the calls not type-safe.

**Comment by lpil** [comment]
Hello! Could you expand on what real-world problems and pain points this feature would solve for Gleam users please? The concrete benefits of adding this feature are not very clear to me yet. Thanks!

**Reply by Shad0wRim** [reply]
I want to create a custom number type, but am unable to make it type safe without hard coding values into it.

```gleam
pub type QRootN(const n: Int) {
  QRootN(int: Int, root: Int, denom: Int)
}
```
These are numbers of the form (a + b*sqrt(n)) / c, and currently the only way I can see to make these type safe across various n values would be to hard code a different type per radicand, having a value in the constructor to denote the radicand would not be found at compile time. By allowing a value type parameter, this could allow these different number types to be checked by the type checker.

Another example would be for a data structure of a specific size.

```gleam
pub opaque type BTree(a, const n: Int) {
  Node(value: a, depth: Int, right: BTree(a), left: BTree(a))
  Leaf(value: a, depth: Int)
}

fn add_item(tree: BTree(a, n), value: a) -> Result(BTree(a, n), Nil) {
  todo
}

fn increase_max_depth(tree: BTree(a, n), value: a) -> Result(BTree(a, n+1), Nil) {
   todo
}
```

In this case allowing const generics could allow for these differing sized types to be checked at compile time.

**Reply by lpil** [reply]
Why wouldn't you have the size held in the record itself? Adding a lot of complexity to the type system to save 1 int's worth of memory per instance isn't very compelling in these examples.

**Reply by Shad0wRim** [reply]
It's less about the memory saving and more about restricting the APIs to only have certain const parameters, since different values in the type constitute a different type that can be checked, such as a function that will only take in two numbers with the same sqrt(n) or a 5 depth tree. By encoding these values as their own "types" in a way, you don't have to do error checking on these cases to ensure it has the correct value and avoid having Result(_,Nil) when we know that the only way it would fail is if the input value didn't match the structure.

If there is some alternative solution to solve these issues in current gleam I would like to hear it, I had thought of encoding types for each desired radicand in the number example, but couldn't think of a way to associate the type with a value that I want to use in functions.

**Reply by lpil** [reply]
Sorry, I don't see when this would enable something you can't do with a normal value. Could you give an example please 🙏

**Reply by Shad0wRim** [reply]
Currently, if someone wants to use something like bounded integer type, error checking has to be done at runtime in order to validate that the value is actually bounded.
```gleam
type BoundedInt {
  BoundedInt(value: Int, low: Int, high: Int)
}

fn do_stuff(val: BoundedInt) -> Result(a, Nil) {
  case val.value >= val.low && val.value <= val.high {
    True -> Ok(more_stuff(val))
    False -> Err(Nil)
  }
}
```
With a const parameter, this check can be eliminated by knowing that the type enforces the range it can be bounded, for example:
```gleam
opaque type BoundedInt(const low: Int, const high: Int) {
   BoundedInt(val)
}

fn do_stuff(val: BoundedInt(0, 100)) -> a {
  more_stuff(val)
}
```
The function signature knows that due to the restrictions of the type, the value must be in the correct range and this error checking can be solely done by the constructor of the type, while also removing the error types from the return since we don't have to validate input, the type validates it.

Another example would be with a mathematical field, such as Z[√n], represented by (a + b√n)
```gleam
type ZRootN {
  ZRootN(a: Int, b: Int, n: Int)
}

fn add(x: ZRootN, y: ZRootN) -> Result(ZRootN, Nil) {
  case x.n == y.n {
    True -> ZRootN(x.a + y.a, x.b + y.b, x.n)
    False -> Err(Nil)
  }
}
```
The fields Z[√2] and Z[√3] are separate spaces and shouldn't interact with one another, and this can be checked at compile time while still being general using a const parameter:
```gleam
type ZRootN(const n: Int) {
  ZRootN(a: Int, b: Int)
}

fn add(x: ZRootN(n), y: ZRootN(n)) -> ZRootN(n) {
  ZRootN(x.a + y.a, x.b + y.b)
}
```
This removes the need for any error checking done, removing the Result from the signature and making add make more sense (adding two numbers should be infallible) and removes code duplication if multiple types of these types want to be used, since each value in the parameter would constitute a separate type.

**Reply by Shad0wRim** [reply]
Another use case would be for using the newtype pattern by not introducing new types, but by grouping them together with one type but using the variants to distinguish the types:
```gleam
type State {
  Valid
  Pending
  Invalid
}
type Process(const s: State) {
  Data(v: Int)
}

fn do_stuff(val: Process(Valid)) {
  todo
}
```
Since the parameter is a value, perhaps it can be returned and acted on by other functions.
```gleam
fn get_state(val: Process(s)) -> State {
  s
}
```

**Reply by lpil** [reply]
What would these phantom type level values give you that regular phantom types do not?

```gleam
pub type N0

pub type N100

pub opaque type BoundedInt(low, high) {
   BoundedInt(val)
}

pub fn do_stuff(val: BoundedInt(N0, N100)) -> a {
  more_stuff(val)
}
```

**Reply by Shad0wRim** [reply]
Being able to associate a value with the type would allow the constructor for this type to be simultaneously generalized and type safe, If there is a way to associate a value with a type in a way that allows this without this feature then this feature would be redundant, but I couldn't find a way to do that.
```gleam
pub opaque type BoundedInt(const low: Int, const high: Int) {
  BoundedInt(val)
}

pub fn new(val, low, high) -> Result(BoundedInt(low, high), Nil) {
  case val <= high && val >= low {
    True -> Ok(BoundedInt(val))
    False -> Err(Nil)
  }
}
```

The way to do this with regular phantom types the way I see it would be to create a separate function for each one:
```gleam
pub type N100
pub type N50
pub type N0

pub opaque type BoundedInt(low, high) {
  BoundedInt(val)
}

pub fn new_0_100(val) -> Result(BoundedInt(N0, N100), Nil) {
  case val <= 100 && val >= 0 {
    True -> Ok(BoundedInt(val))
    False -> Err(Nil)
  }
}

pub fn new_0_50(val) -> Result(BoundedInt(N0, N100), Nil {
  todo
}
```

This creates a lot of mental overhead for adding simply one for value for the bounds that this type can have. If there is a way to do this in gleam already without this, then I would absolutely use that instead.

**Reply by lpil** [reply]
That example you've given there requires a lot more than the feature you've specified in the first post, rather it would require dependant types.

Dependant types are an active area of computer science research and it is not yet understood how they can be practically utilised in a programming language.

This is an extremely complex area of research and I don't believe we would be able to do anything in this area.
